File: Source/UI/Test Runner/IPFTestRunnerViewController.m
#import "IPFTestRunnerViewController.h"
#import "IPFTestRunner.h"
#import "IPFTestRunnerConfiguration.h"
#import "IPFIcon.h"
#import "IPFHelpViewController.h"
#import "iperf-Swift.h"
@import AudioToolbox;

static int getTestDuration(NSUInteger selectedSegmentIndex)
{
    switch (selectedSegmentIndex) {
        case 0:
            return 10;

        case 1:
            return 30;

        case 2:
            return 300;

        default:
            return 10;
    }
}

@interface IPFTestRunnerViewController ()

@property (strong, nonatomic) IPFTestRunner *testRunner;

@end

@implementation IPFTestRunnerViewController {
    CGFloat _averageBandwidthTotal;
    NSUInteger _averageBandwidthCount;
    CGFloat _maxBandwidth;
    CGFloat _minBandwidth;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];

    if (self != nil) {
        self.title = NSLocalizedString(@"TestRunner.title", @"Main screen title");
        self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"TestRunner.Help", @"Help start button name")
                                                                                 style:UIBarButtonItemStylePlain
                                                                                target:self
                                                                                action:@selector(showHelp)];
        [self showStartButton:YES];
    }

    return self;
}

- (IBAction)showHelp
{
    IPFHelpViewController *helpViewController = [[IPFHelpViewController alloc] initWithNibName:nil bundle:nil];

    [self.navigationController pushViewController:helpViewController animated:YES];
}

- (void)showStartButton:(BOOL)showStartButton
{
    NSString *title = showStartButton ? NSLocalizedString(@"TestRunner.start", @"Test start button name") : NSLocalizedString(@"TestRunner.stop", @"Test stop button name");
    UIBarButtonItem *buttonItem = [[UIBarButtonItem alloc] initWithTitle:title
                                                                   style:UIBarButtonItemStylePlain
                                                                  target:self
                                                                  action:@selector(startStopTest)];

    if (!showStartButton) {
        buttonItem.tintColor = [UIColor redColor];
    }

    self.navigationItem.rightBarButtonItem = buttonItem;
}

- (void)viewDidLoad
{
    [self restoreTestSettings];
}

- (void)restoreTestSettings
{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *hostname = [defaults stringForKey:@"IPFTestHostname"];
    NSNumber *port = [defaults objectForKey:@"IPFTestPort"];

    if ([hostname length] > 0 && [port unsignedIntegerValue] > 0) {
        self.addressTextField.text = hostname;
        self.portTextField.text = [port stringValue];
    }
}

- (void)saveTestSettings
{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *hostname = self.addressTextField.text;
    NSNumber *port = [NSNumber numberWithUnsignedInteger:(NSUInteger)[self.portTextField.text integerValue]];

    if ([hostname length] > 0 && [port unsignedIntegerValue] > 0) {
        [defaults setObject:hostname forKey:@"IPFTestHostname"];
        [defaults setObject:port forKey:@"IPFTestPort"];
        [defaults synchronize];
    }
}

- (void)saveTestResults
{
    IPFTestResult *testResult = [[IPFTestResult alloc] init];
    testResult.date = [NSDate new];
    testResult.mode = (self.transmitModeSlider.selectedSegmentIndex == 0) ? @"⇈" : @"⇊";
    testResult.duration = getTestDuration(self.testDurationSlider.selectedSegmentIndex);
    testResult.streams = self.streamsSlider.selectedSegmentIndex + 1;
    testResult.averageBandWidth = 123;
    testResult.location = self.locationTextField.text;

    [[IPFTestResultsManager shared] add:testResult];
    AudioServicesPlaySystemSound(1109);
    UINotificationFeedbackGenerator *gen = [UINotificationFeedbackGenerator new];
    [gen notificationOccurred:UINotificationFeedbackTypeSuccess];
}

- (void)startStopTest
{
    IPFTestRunner *testRunner = self.testRunner;

    if (testRunner) {
        [testRunner stopTest];
    } else {
        [self startTest];
        [self.view endEditing:true];
    }
}

- (void)startTest
{
    IPFTestRunnerConfiguration *configuration = [[IPFTestRunnerConfiguration alloc] initWithHostname:self.addressTextField.text
                                                                                                port:[self.portTextField.text intValue]
                                                                                            duration:getTestDuration(self.testDurationSlider.selectedSegmentIndex)
                                                                                             streams:[self.streamsSlider selectedSegmentIndex] + 1
                                                                                                type:[self.transmitModeSlider selectedSegmentIndex]];
    IPFTestRunner *testRunner = [[IPFTestRunner alloc] initWithConfiguration:configuration];
    UIApplication *application = [UIApplication sharedApplication];
    __block UIBackgroundTaskIdentifier backgroundTask = [application beginBackgroundTaskWithExpirationHandler:^{
        [application endBackgroundTask:backgroundTask];
        backgroundTask = UIBackgroundTaskInvalid;
    }];

    self.testRunner = testRunner;
    [self showStartButton:NO];
    self.addressTextField.enabled = NO;
    self.portTextField.enabled = NO;
    self.transmitModeSlider.enabled = NO;
    self.streamsSlider.enabled = NO;
    self.testDurationSlider.enabled = NO;
    self.bandwidthLabel.text = @"...";
    self.averageBandwidthLabel.text = @"";
    self.progressView.progress = 0.0;
    self.progressView.hidden = YES;
    [application setNetworkActivityIndicatorVisible:YES];
    _averageBandwidthTotal = 0;
    _averageBandwidthCount = 0;
    _maxBandwidth = CGFLOAT_MIN;
    _minBandwidth = CGFLOAT_MAX;

    [testRunner startTest:^(IPFTestRunnerStatus status) {
        switch (status.errorState) {
            case IPFTestRunnerErrorStateNoError:
                break;

            case IPFTestRunnerErrorStateCouldntInitializeTest:
                [self showAlert:NSLocalizedString(@"TestRunner.errorInitializing", "Error initializing the test")];
                break;

            case IPFTestRunnerErrorStateCannotConnectToTheServer:
                [self showAlert:NSLocalizedString(@"TestRunner.connectionError", @"Cannot connect to the server, please check that the server is running")];
                break;

            case IPFTestRunnerErrorStateServerIsBusy:
                [self showAlert:NSLocalizedString(@"TestRunner.serverBusy", @"Server is busy, please retry later")];
                break;

            default:
                [self showAlert:[NSString stringWithFormat:NSLocalizedString(@"TestRunner.errorUnknown", @"Unknown error %d running the test"), status.errorState]];
                break;
        }

        if (status.errorState != IPFTestRunnerErrorStateNoError) {
            [self showAlert:NSLocalizedString(@"TestRunner.errorDefault", @"Error running the test")];
        }

        if (status.running == NO) {
            [self showStartButton:YES];
            self.addressTextField.enabled = YES;
            self.portTextField.enabled = YES;
            self.transmitModeSlider.enabled = YES;
            self.streamsSlider.enabled = YES;
            self.testDurationSlider.enabled = YES;
            self.progressView.hidden = YES;
            self.testRunner = nil;
            [application setNetworkActivityIndicatorVisible:NO];
            [application endBackgroundTask:backgroundTask];

            if (status.errorState == IPFTestRunnerErrorStateNoError || status.errorState == IPFTestRunnerErrorStateServerIsBusy) {
                if (self->_averageBandwidthTotal) {
                    self.bandwidthLabel.text = [NSString stringWithFormat:@"%.0f Mbits/s", self->_averageBandwidthTotal / (CGFloat)self->_averageBandwidthCount];
                    self.averageBandwidthLabel.text = [NSString stringWithFormat:@"min: %.0f max: %.0f", self->_minBandwidth, self->_maxBandwidth];
                    self.progressView.hidden = NO;
                } else {
                    self.bandwidthLabel.text = @"";
                }

                // Only persist settings if the test is successful
                [self saveTestSettings];
                // Save results
                [self saveTestResults];
            }
        } else {
            CGFloat bandwidth = status.bandwidth;

            self->_averageBandwidthTotal += status.bandwidth;
            self->_averageBandwidthCount++;
            self.progressView.hidden = NO;

            if (bandwidth < self->_minBandwidth) {
                self->_minBandwidth = bandwidth;
            }

            if (bandwidth > self->_maxBandwidth) {
                self->_maxBandwidth = bandwidth;
            }

            self.bandwidthLabel.text = [NSString stringWithFormat:@"%.0f Mbits/s", status.bandwidth];
            self.averageBandwidthLabel.text = [NSString stringWithFormat:@"avg: %.0f min: %.0f max: %.0f", self->_averageBandwidthTotal / (CGFloat)self->_averageBandwidthCount, self->_minBandwidth, self->_maxBandwidth];
            [self.progressView setProgress:status.progress animated:YES];
        }
    }];
}

- (void)showAlert:(NSString *)alertText
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:alertText message:nil preferredStyle:UIAlertControllerStyleAlert];

    [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK", @"OK Action") style:UIAlertActionStyleDefault handler:NULL]];
    [self presentViewController:alertController animated:YES completion:NULL];
    self.bandwidthLabel.text = @"";
    self.averageBandwidthLabel.text = @"";
    self.progressView.hidden = YES;
}

@end

File: Source/UI/Test Runner/IPFTestRunnerViewController.swift
import AudioToolbox
import UIKit

private enum IPFTestState {
    case active
    case inactive
}

final class IPFTestRunnerViewController: FormViewController {
    override func viewDidLoad() {
        super.viewDidLoad()
        restoreTestSettings()
        setupUI()
        updateUI(state: .inactive)
    }

    private var configuration = IPFTestRunnerConfiguration()
    private var testRunner: IPFTestRunner?
    private var testLocation = ""
    private let durations: [UInt] = [5, 10, 30, 300]
    private var resultsHeader = IPFTestResultsHeaderView()
    private lazy var progressView = UIProgressView(progressViewStyle: .bar)
    private var speedTestActiveTaskIdentifier: UIBackgroundTaskIdentifier = .invalid
}

private extension IPFTestRunnerViewController {
    // MARK: - User Actions

    @objc func showHelp() {
        show(IPFHelpViewController(), sender: self)
    }

    @objc func startStopTest() {
        if configuration.hostname.isEmpty || !(0...65535).contains(configuration.port) {
            showAlert(message: "TestRunner.errorEmptyFields".localized)
            return
        }

        if testRunner != nil {
            testRunner?.stopTest()
        } else {
            resultsHeader.showInitial()
            startTest()
        }
    }

    // MARK: - Other Private methods

    func setupUI() {
        title = "TestRunner.title".localized

        if #available(iOS 13.0, *) {
            view.backgroundColor = .systemBackground
        } else {
            view.backgroundColor = .white
        }

        navigationItem.leftBarButtonItem = UIBarButtonItem(title: "TestRunner.Help".localized,
                                                           style: .plain,
                                                           target: self, action: #selector(showHelp))

        if let navBar = navigationController?.navigationBar {
            navigationController?.navigationBar.addSubview(progressView)
            progressView.translatesAutoresizingMaskIntoConstraints = false
            NSLayoutConstraint.activate([
                progressView.leftAnchor.constraint(equalTo: navBar.leftAnchor),
                progressView.rightAnchor.constraint(equalTo: navBar.rightAnchor),
                progressView.bottomAnchor.constraint(equalTo: navBar.bottomAnchor),
            ])
        }

        // Setup the individual UI fields

        form.addSection(Section())

        let header = ViewRow {
            $0.childView = resultsHeader
        }
        form.addRow(header)

        let serverAddress = TextFieldRow {
            $0.placeHolder = "TestRunner.serverAddress".localized
            $0.keyboardType = .numbersAndPunctuation
            $0.text = configuration.hostname
        }.onTextChange { [weak self] text in
            self?.configuration.hostname = text
        }
        form.addRow(serverAddress)

        let serverPort = TextFieldRow {
            $0.placeHolder = "TestRunner.serverPort".localized
            $0.keyboardType = .numberPad
            $0.text = String(configuration.port)
        }.onTextChange { [weak self] text in
            self?.configuration.port = UInt(text) ?? 0
        }
        form.addRow(serverPort)

        let testLocation = TextFieldRow {
            $0.placeHolder = "TestRunner.testLocation".localized
            $0.keyboardType = .default
        }.onTextChange { [weak self] text in
            self?.testLocation = text
        }
        form.addRow(testLocation)

        form.addSection(Section())

        let transmitMode = SegmentedRow {
            $0.text = "TestRunner.transmitMode".localized
            $0.segments = ["Upload", "Download"]
            $0.defaultSegment = Int(configuration.type.rawValue)
        }.onSegmentChange { [weak self] selectedIndex, _ in
            self?.configuration.type = IPFTestRunnerConfigurationType(rawValue: UInt(selectedIndex)) ?? .upload
        }
        form.addRow(transmitMode)

        let streams = SegmentedRow {
            $0.text = "TestRunner.streams".localized
            $0.segments = (1 ... 5).compactMap { String($0) }
            $0.defaultSegment = Int(configuration.streams) - 1
        }.onSegmentChange { [weak self] selectedIndex, _ in
            self?.configuration.streams = UInt(selectedIndex) + 1
        }
        form.addRow(streams)

        let duration = SegmentedRow {
            $0.text = "TestRunner.duration".localized
            $0.segments = durations.compactMap { seconds -> String in
                seconds < 60 ? "\(seconds)s" : "\(seconds / 60) min"
            }
            $0.defaultSegment = durations.firstIndex(of: configuration.duration) ?? 3
        }.onSegmentChange { [weak self] selectedIndex, _ in
            self?.configuration.duration = self?.durations[selectedIndex] ?? 10
        }
        form.addRow(duration)
    }

    func restoreTestSettings() {
        configuration.hostname = UserDefaults.hostname
        configuration.port = UserDefaults.port
        configuration.type = IPFTestRunnerConfigurationType(rawValue: UserDefaults.configType) ?? .upload
        configuration.streams = UserDefaults.streams
        configuration.duration = UserDefaults.duration
    }

    func saveTestSettings() {
        UserDefaults.hostname = configuration.hostname
        UserDefaults.port = configuration.port
        UserDefaults.configType = configuration.type.rawValue
        UserDefaults.streams = configuration.streams
        UserDefaults.duration = configuration.duration
    }

    func saveTestResults() {
        let testResult = IPFTestResult()
        testResult.date = Date()
        testResult.mode = configuration.type == .download ? "⇊" : "⇈"
        testResult.duration = configuration.duration
        testResult.streams = configuration.streams
        testResult.averageBandWidth = resultsHeader.averageBandWidth
        testResult.location = testLocation
        IPFTestResultsManager.shared.add(testResult)
    }

    func playFeedBack() {
        if UserDefaults.enableSounds {
            // A nice sound effect built-in to iOS
            AudioServicesPlayAlertSound(1109)
        }

        if #available(iOS 10.0, *), UserDefaults.enableHaptics {
            // Play a success notification haptic
            let feedbackGenerator = UINotificationFeedbackGenerator()
            feedbackGenerator.notificationOccurred(.success)
        }
    }

    func updateUI(state: IPFTestState) {
        let actionTitle: String

        switch state {
        case .active:
            actionTitle = "TestRunner.stop"
            form.enabled = false
            progressView.progress = 0
            progressView.isHidden = false
            UIApplication.shared.isNetworkActivityIndicatorVisible = true

        case .inactive:
            actionTitle = "TestRunner.start"
            form.enabled = true
            progressView.progress = 0
            progressView.isHidden = true
            UIApplication.shared.isNetworkActivityIndicatorVisible = false
        }

        view.endEditing(true)

        let rightButton = UIBarButtonItem(title: actionTitle.localized, style: .plain, target: self, action: #selector(startStopTest))
        if state == .active {
            rightButton.tintColor = .red
        }
        navigationItem.rightBarButtonItem = rightButton
    }

    func startTest() {
        updateUI(state: .active)

        let app = UIApplication.shared
        speedTestActiveTaskIdentifier = app.beginBackgroundTask(expirationHandler: { [weak self] in
            guard let self = self else { return }
            app.endBackgroundTask(self.speedTestActiveTaskIdentifier)
        })

        testRunner = IPFTestRunner(configuration: configuration)
        testRunner?.startTest { [weak self] status in
            guard let self = self else { return }

            switch status.errorState {
            case .noError:
                break
            case .couldntInitializeTest:
                self.showAlert(message: "TestRunner.errorInitializing".localized)
            case .cannotConnectToTheServer:
                self.showAlert(message: "TestRunner.errorConnection".localized)
            case .serverIsBusy:
                self.showAlert(message: "TestRunner.errorServerBusy".localized)
            case .unknown:
                self.showAlert(message: "TestRunner.errorUnknown".localizeWithFormat(arguments: status.errorState.rawValue))
            @unknown default:
                break
            }

            if status.running.boolValue == false {
                self.resultsHeader.showFinal()
                app.endBackgroundTask(self.speedTestActiveTaskIdentifier)
                self.updateUI(state: .inactive)
                self.testRunner = nil
                if status.errorState == .noError || status.errorState == .serverIsBusy {
                    if self.resultsHeader.averageBandWidth > 0 {
                        self.saveTestResults()
                        self.saveTestSettings()
                        self.playFeedBack()
                    }
                }
            } else {
                self.updateUI(state: .active)
                self.progressView.setProgress(Float(status.progress), animated: true)
                self.resultsHeader.currentBandWidth = Int(status.bandwidth)
            }
        }
    }

    func showAlert(message: String) {
        resultsHeader.reset()

        let alert = UIAlertController(title: "Warning".localized, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK".localized, style: .default, handler: nil))
        present(alert, animated: true, completion: nil)
    }
}
